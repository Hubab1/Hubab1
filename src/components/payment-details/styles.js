import { css } from 'emotion';

export const cardHeader = css`
    align-self: start;
    flex: 1;
    padding-left: 0;
    font-weight: 400;
    display: block;
    font-size: 14px;
`;

export const headerContainer = css`
    display: flex;
    flex: 1;
    align-items: center;
    text-transform: none;
    padding-bottom: 15px;
`;

export const imageWrapper = css`
    line-height: 67px;
    width: 42px;
    text-align: left;
    margin: 0;
    padding: 0;
    margin-right: 10px;
`;

export const itemsContainer = css`
    margin-bottom: 0;
    border-bottom: 1px solid #eeeeee;

    .MuiExpansionPanelSummary-root {
        padding: 0;
    }

    .MuiExpansionPanelSummary-root.Mui-expanded {
        min-height: 20px;
        height: 45px;
        margin: 0;
    }

    :nth-child(2) {
        border-top: 1px solid #eeeeee;
    }
`;

export const itemRow = css`
    padding: 10px 0;
    text-align: left;
    border-top: 1px solid #eeeeee;
    font-size: 15px;
`;

export const paidText = css`
    position: absolute;
    right: 28px;
    top: 13px;
`;

export const paymentDetailRow = css`
    display: flex;
    justify-content: space-between;
    padding-bottom: 9px;
    font-size: 14px;

    div:first-child {
        padding-right: 50px;
    }
`;

export const paymentDetailTotal = css`
    display: flex;
    justify-content: space-between;
    background-color: rgba(103, 193, 139, 0.1);
    padding: 10px 15px 10px 16px;
    width: 458px;
    margin-left: -15px;
    font-size: 15px;

    div:first-child {
        font-weight: bold;
    }
`;

export const title = css`
    font-weight: bold;
`;

export const rentalOptionName = css`
    padding-right: 20px;
`;

export const styles = {
    cardHeader,
    headerContainer,
    itemsContainer,
    itemRow,
    imageWrapper,
    paidText,
    paymentDetailRow,
    paymentDetailTotal,
    title,
    rentalOptionName,
};

export const viewPaymentTerms = css`
    color: #2b44ff;
    cursor: pointer;
    text-decoration: underline;
    padding-left: 5px;
`;

export const termsDiv = css`
    margin-bottom: 20px;
    margin-top: 44px;
    font-size: 14px;
`;
