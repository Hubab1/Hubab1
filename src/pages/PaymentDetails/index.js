import PaymentDetailsPage from './components/PaymentDetailsPage/PaymentDetailsPage';

export default PaymentDetailsPage;
