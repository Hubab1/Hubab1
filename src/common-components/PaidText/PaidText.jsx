import React from 'react';
import PropTypes from 'prop-types';
import GreenCheckIcon from 'common-components/GreenCheckIcon/GreenCheckIcon';
import { P } from 'assets/styles';

export const PaidText = ({ margin = '0' }) => (
    <P bold color="#56BA82" margin={margin}>
        <GreenCheckIcon />
        Paid
    </P>
);

PaidText.propTypes = {
    margin: PropTypes.string,
};

export default PaidText;
