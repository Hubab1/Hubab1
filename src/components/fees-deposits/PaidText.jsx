import React from 'react';

import GreenCheckIcon from 'components/common/GreenCheckIcon';
import { P } from 'assets/styles';


export const PaidText = ({margin="0"}) => <P bold color="#56BA82" margin={margin}><GreenCheckIcon/>Paid</P>;

export default PaidText;