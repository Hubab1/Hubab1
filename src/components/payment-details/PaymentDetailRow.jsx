import PropTypes from 'prop-types';
import React from 'react';
import { styles } from './styles';
import { getPaymentItemName } from 'utils/misc';

export const TYPE_CONCESSION = 'concession';

export function PaymentDetailRows({ paymentObject, paymentType }) {
    const rows = paymentObject.items.map((item) => (
        <PaymentDetailRow
            key={`${item.name}-${paymentType}`}
            className={styles.paymentDetailRow}
            name={item.name}
            paymentTotal={item.amount}
            quantity={item.quantity}
            price={item.price}
            included={item.included}
            perDay={item.per_day}
            prorated={item.prorated}
            days={item.days}
            exempted={item.exempted}
            type={item.type}
        />
    ));

    rows.push(
        <PaymentDetailRow
            key={`total${paymentType}`}
            name="Total"
            paymentTotal={paymentObject.total}
            className={styles.paymentDetailTotal}
        />
    );

    return rows;
}

PaymentDetailRows.propTypes = {
    paymentObject: PropTypes.object.isRequired,
    paymentType: PropTypes.string.isRequired,
};

export const PaymentDetailRow = ({
    name,
    paymentTotal,
    quantity = 1,
    price,
    included = 0,
    className,
    prorated,
    days,
    perDay,
    exempted,
    type,
}) => {
    if (!quantity) return null;

    return (
        <div className={className}>
            <div>
                {getRentalOptionFeeDisplayName(name, price, quantity, included, prorated, days, perDay, exempted)}
            </div>
            <div>
                {type === TYPE_CONCESSION && `-`}${paymentTotal}
            </div>
        </div>
    );
};

PaymentDetailRow.propTypes = {
    className: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    paymentTotal: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    quantity: PropTypes.number,
    price: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    included: PropTypes.number,
    prorated: PropTypes.bool,
    days: PropTypes.number,
    perDay: PropTypes.string,
    exempted: PropTypes.number,
    type: PropTypes.string,
};

// This should match the logic in chuck
export function getRentalOptionFeeDisplayName(
    name,
    price,
    quantity = 1,
    included = 0,
    prorated,
    days,
    perDay,
    exempted = 0
) {
    if (quantity < 1) return null;

    const feesBreakdown = [];
    const num_included = quantity > included ? included : quantity;
    if (num_included > 0) {
        feesBreakdown.push(`${num_included} incl.`);
    }

    if (exempted) {
        feesBreakdown.push(`${exempted} service animal${exempted > 1 ? 's' : ''} incl.`);
    }

    const uncharged_quantity = included + exempted;
    const num_charged = quantity > uncharged_quantity ? quantity - uncharged_quantity : 0;
    if ((num_charged > 0 || exempted > 0) && quantity > 1 && !prorated) {
        feesBreakdown.push(`${num_charged} x $${price}`);
    }

    let displayName = `${getPaymentItemName(name)}`;
    const daysPlural = days > 1 ? 's' : '';

    if (prorated && num_charged > 0 && quantity > 0) {
        displayName += `: ${days} day${daysPlural} at $${perDay}/day`;
    }

    if (feesBreakdown.length > 0) {
        displayName += ` (${feesBreakdown.join(', ')})`;
    }

    return displayName;
}
