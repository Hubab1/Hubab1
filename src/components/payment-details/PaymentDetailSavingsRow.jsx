import { styles } from 'components/payment-details/styles';
import PropTypes from 'prop-types';
import React from 'react';

export function PaymentDetailSavingsRows({ paymentObject, paymentType }) {
    const rows = paymentObject.items.map((item) => (
        <PaymentDetailSavingsRow
            key={`${item.name}-${paymentType}`}
            name={item.name}
            paymentTotal={item.amount}
            pricePerMonth={item.price}
            months={item.months}
            className={styles.paymentDetailRow}
        />
    ));
    rows.push(
        <PaymentDetailSavingsRow
            key={`total${paymentType}`}
            name={'Total'}
            paymentTotal={paymentObject.total}
            className={styles.paymentDetailTotal}
        />
    );
    return rows;
}

PaymentDetailSavingsRows.propTypes = {
    paymentObject: PropTypes.object.isRequired,
    paymentType: PropTypes.string.isRequired,
};

export const PaymentDetailSavingsRow = ({ paymentTotal, name, months, pricePerMonth, className }) => {
    if (name.toLowerCase() === 'rentable item concession') {
        name = 'Parking, Storage, Other Monthly Charge Concession';
    }

    return (
        <div className={className}>
            <div>
                {name}
                {months && `: ${months} months at $${pricePerMonth}/month`}
            </div>
            <div>${paymentTotal}</div>
        </div>
    );
};

PaymentDetailSavingsRow.propTypes = {
    name: PropTypes.string.isRequired,
    paymentTotal: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    months: PropTypes.number,
    pricePerMonth: PropTypes.string,
};
