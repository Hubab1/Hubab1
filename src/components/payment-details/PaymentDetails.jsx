import React from 'react';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { ROUTES } from 'app/constants';
import { Card, H1, SpacedH3 } from 'assets/styles';
import captureRoute from 'app/captureRoute';
import lightbulb from 'assets/images/lightbulb.png';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';

import { makeStyles } from '@material-ui/core/styles';
import { PaymentDetailRows } from 'components/payment-details/PaymentDetailRow';
import { styles } from './styles';
import { format, parseISO } from 'date-fns';

const useStyles = makeStyles(() => ({
    root: {
        display: 'block',
        padding: 0,
    },
}));

export const PaymentCard = styled(Card)`
    padding: 15px;
`;

export const PaymentDetails = ({ profile, configuration }) => {
    if (!profile || !configuration || !profile.lease_start_date) return null;

    const leaseStartDate = format(parseISO(profile.lease_start_date), 'MM/dd/yyyy');

    return (
        <>
            <H1>Payment Details</H1>
            <SpacedH3>Hereâ€™s the breakdown for unit {profile.unit.unit_number}</SpacedH3>
            <PaymentCard>
                <div className={styles.headerContainer}>
                    <div className={styles.imageWrapper}>
                        <img alt="coin" src={lightbulb} />
                    </div>
                    <div className={styles.cardHeader}>
                        <div className={styles.title}>Payment breakdown</div>
                        <div>Move in date: {leaseStartDate}</div>
                        <div>Lease term: {profile.lease_term} months</div>
                    </div>
                </div>
                <PaymentItemsExpansionPanel
                    amount={profile.fees_breakdown.application_fees.total}
                    label={'due at application'}
                    defaultExpanded={false}
                >
                    <PaymentDetailRows
                        paymentObject={profile.fees_breakdown.application_fees}
                        paymentType="application"
                    />
                </PaymentItemsExpansionPanel>
                <PaymentItemsExpansionPanel
                    amount={profile.fees_breakdown.move_in_fees.total}
                    label={'due at move in'}
                    defaultExpanded={false}
                >
                    <PaymentDetailRows paymentObject={profile.fees_breakdown.move_in_fees} paymentType="move_in" />
                </PaymentItemsExpansionPanel>
                <PaymentItemsExpansionPanel
                    amount={profile.fees_breakdown.monthly_fees.total}
                    label={'monthly rent'}
                    defaultExpanded={false}
                >
                    <PaymentDetailRows paymentObject={profile.fees_breakdown.monthly_fees} paymentType="monthly_fees" />
                </PaymentItemsExpansionPanel>
            </PaymentCard>
        </>
    );
};

PaymentDetails.propTypes = {
    profile: PropTypes.object.isRequired,
    configuration: PropTypes.object.isRequired,
};

function PaymentItemsExpansionPanel({ children, defaultExpanded, label, amount }) {
    const classes = useStyles();

    return (
        <div className={styles.itemsContainer}>
            <ExpansionPanel elevation={0} defaultExpanded={defaultExpanded}>
                <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
                    <b>${amount}</b>&nbsp;{label}
                </ExpansionPanelSummary>
                <ExpansionPanelDetails classes={{ root: classes.root }}>
                    {React.Children.map(children, (child) => (
                        <div className={styles.itemRow}>{child}</div>
                    ))}
                </ExpansionPanelDetails>
            </ExpansionPanel>
        </div>
    );
}

PaymentItemsExpansionPanel.propTypes = {
    children: PropTypes.object,
    defaultExpanded: PropTypes.bool,
    label: PropTypes.string.isRequired,
    amount: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
    profile: state.renterProfile,
    configuration: state.configuration,
});

export default connect(mapStateToProps)(captureRoute(PaymentDetails, ROUTES.PAYMENT_DETAILS));
