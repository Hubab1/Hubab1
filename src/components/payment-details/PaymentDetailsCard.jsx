import { styles } from 'components/payment-details/styles';
import styled from '@emotion/styled';
import lightbulb from 'assets/images/lightbulb.png';
import { PaymentDetailRows } from 'components/payment-details/PaymentDetailRow';
import React from 'react';
import { format, parseISO } from 'date-fns';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import PaidText from 'components/common/PaidText';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import { Card } from 'assets/styles';

const useStyles = makeStyles(() => ({
    root: {
        display: 'block',
        padding: 0,
    },
}));

export const PaymentCard = styled(Card)`
    padding: 15px;
`;

export const PaymentDetailsCard = ({ profile, payables = [] }) => {
    const leaseStartDate = format(parseISO(profile.lease_start_date), 'MM/dd/yyyy');
    const allPaymentsPaid = !payables.find((payment) => !payment.paid);

    return (
        <PaymentCard>
            <div className={styles.headerContainer}>
                <div className={styles.imageWrapper}>
                    <img alt="coin" src={lightbulb} />
                </div>
                <div className={styles.cardHeader}>
                    <div className={styles.title}>Payment breakdown</div>
                    <div>Move in date: {leaseStartDate}</div>
                    <div>Lease term: {profile.lease_term} months</div>
                </div>
            </div>
            <PaymentItemsExpansionPanel
                amount={profile.fees_breakdown.application_fees.total}
                label={'due at application'}
                defaultExpanded={false}
                isPaid={allPaymentsPaid}
            >
                <PaymentDetailRows paymentObject={profile.fees_breakdown.application_fees} paymentType="application" />
            </PaymentItemsExpansionPanel>
            <PaymentItemsExpansionPanel
                amount={profile.fees_breakdown.move_in_fees_v2.total}
                label={'due at move in'}
                defaultExpanded={false}
            >
                <PaymentDetailRows paymentObject={profile.fees_breakdown.move_in_fees_v2} paymentType="move_in" />
            </PaymentItemsExpansionPanel>
            <PaymentItemsExpansionPanel
                amount={profile.fees_breakdown.monthly_fees.total}
                label={'monthly rent'}
                defaultExpanded={false}
            >
                <PaymentDetailRows paymentObject={profile.fees_breakdown.monthly_fees} paymentType="monthly_fees" />
            </PaymentItemsExpansionPanel>
        </PaymentCard>
    );
};

PaymentDetailsCard.propTypes = {
    profile: PropTypes.object.isRequired,
    payables: PropTypes.array,
};

export function PaymentItemsExpansionPanel({ children, defaultExpanded, label, amount, isPaid }) {
    const classes = useStyles();

    return (
        <div className={styles.itemsContainer}>
            <ExpansionPanel elevation={0} defaultExpanded={defaultExpanded}>
                <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
                    <div className={styles.itemsContainerLabel}>
                        <b>${amount}</b>&nbsp;{label}
                    </div>
                    {isPaid && (
                        <div className={styles.paidText}>
                            <PaidText />
                        </div>
                    )}
                </ExpansionPanelSummary>
                <ExpansionPanelDetails classes={{ root: classes.root }}>
                    {React.Children.map(children, (child) => (
                        <div className={styles.itemRow}>{child}</div>
                    ))}
                </ExpansionPanelDetails>
            </ExpansionPanel>
        </div>
    );
}

PaymentItemsExpansionPanel.propTypes = {
    children: PropTypes.object,
    defaultExpanded: PropTypes.bool,
    label: PropTypes.string.isRequired,
    amount: PropTypes.string.isRequired,
    isPaid: PropTypes.bool,
};
