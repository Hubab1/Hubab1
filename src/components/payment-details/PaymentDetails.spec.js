import { shallow, mount } from 'enzyme';
import { PaymentDetails } from 'components/payment-details/PaymentDetails';
import React from 'react';
import PaidText from 'components/common/PaidText';

describe('PaymentDetails', () => {
    let defaultProps;

    beforeEach(() => {
        defaultProps = {
            profile: {
                lease_start_date: '2020-12-15',
                unit: {
                    unit_number: 1,
                },
                lease_term: 12,
                fees_breakdown: {
                    application_fees: {
                        total: '620',
                        items: [],
                    },
                    move_in_fees_v2: {
                        total: '0',
                        items: [],
                    },
                    monthly_fees: {
                        total: '1234',
                        items: [],
                    },
                },
            },
            configuration: {},
            payables: [{ paid: true }, { paid: true }, { paid: true }],
            fetchPayments: jest.fn(),
        };
    });

    it('renders the correct number of rows', () => {
        const wrapper = shallow(<PaymentDetails {...defaultProps} />);
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('renders the "Paid" text if all due at application payments have been paid', () => {
        const wrapper = mount(<PaymentDetails {...defaultProps} />);

        expect(defaultProps.fetchPayments).toBeCalledTimes(1);
        expect(wrapper.find(PaidText).length).toEqual(1);
    });

    it('does not render the "Paid" text if some due at application payments have not been paid', () => {
        const payables = [{ paid: true }, { paid: false }, { paid: true }];

        const wrapper = mount(<PaymentDetails {...{ ...defaultProps, payables }} />);

        expect(defaultProps.fetchPayments).toBeCalledTimes(1);
        expect(wrapper.find(PaidText).length).toEqual(0);
    });

    it('displays terms link when passed payment', () => {
        const applicant = {
            receipt: { id: 123 },
        };
        const wrapper = shallow(<PaymentDetails {...defaultProps} applicant={applicant} />);
        expect(wrapper.text()).toContain('Payment and Holding Deposit Terms');
    });

    it('Does not display terms link when payment is incomplete/pending', () => {
        const applicant = {
            receipt: null,
        };
        const wrapper = shallow(<PaymentDetails {...defaultProps} applicant={applicant} />);
        expect(wrapper.text()).not.toContain('Payment and Holding Deposit Terms');
    });
});
