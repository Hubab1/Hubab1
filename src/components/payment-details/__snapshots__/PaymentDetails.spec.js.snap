// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentDetails renders the correct number of rows 1`] = `
<React.Fragment>
  <ForwardRef(render)>
    Payment Details
  </ForwardRef(render)>
  <ForwardRef(render)>
    Hereâ€™s the breakdown for unit 
    1
  </ForwardRef(render)>
  <PaymentDetailsCard
    fetchPayments={[MockFunction]}
    payables={
      Array [
        Object {
          "paid": true,
        },
        Object {
          "paid": true,
        },
        Object {
          "paid": true,
        },
      ]
    }
    profile={
      Object {
        "fees_breakdown": Object {
          "application_fees": Object {
            "items": Array [],
            "total": "620",
          },
          "monthly_fees_v2": Object {
            "items": Array [],
            "total": "1234",
          },
          "move_in_fees_v2": Object {
            "items": Array [],
            "total": "0",
          },
        },
        "lease_start_date": "2020-12-15",
        "lease_term": 12,
        "unit": Object {
          "unit_number": 1,
        },
      }
    }
  />
</React.Fragment>
`;
